tests/drivers/dac/dac_emul/src/main.c

```diff
@@ -0,0 +1,589 @@
+		.channel_id = 0,
+		.resolution = 12,
+	};
+	int ret = dac_channel_setup(dac_emul0, &cfg);
+
+	zassert_equal(ret, 0, "Setup should succeed");
+
+	/* Write a value */
+	uint32_t test_value = 2048; /* Half of 12-bit range */
+
+	ret = dac_write_value(dac_emul0, 0, test_value);
+	zassert_equal(ret, 0, "Write should succeed");
+
+	/* Verify the value was written */
+	uint32_t read_value;
+
+	ret = dac_emul_value_get(dac_emul0, 0, &read_value);
+	zassert_equal(ret, 0, "Read should succeed");
+	zassert_equal(read_value, test_value, "Read value should match written value");
+}
+
+ZTEST(dac_emul_tests, test_write_value_boundary)
+{
+	/* Setup 8-bit channel */
+	struct dac_channel_cfg cfg = {
+		.channel_id = 0,
+		.resolution = 8,
+	};
+	int ret = dac_channel_setup(dac_emul0, &cfg);
+
+	zassert_equal(ret, 0, "Setup should succeed");
+
+	/* Test minimum value (0) */
+	ret = dac_write_value(dac_emul0, 0, 0);
+	zassert_equal(ret, 0, "Write min value should succeed");
+
+	uint32_t value;
+
+	ret = dac_emul_value_get(dac_emul0, 0, &value);
+	zassert_equal(value, 0, "Min value should be 0");
+
+	/* Test maximum value (255 for 8-bit) */
+	ret = dac_write_value(dac_emul0, 0, 255);
+	zassert_equal(ret, 0, "Write max value should succeed");
+
+	ret = dac_emul_value_get(dac_emul0, 0, &value);
+	zassert_equal(value, 255, "Max value should be 255");
+}
+
+ZTEST(dac_emul_tests, test_write_value_out_of_range)
+{
+	/* Setup 8-bit channel */
+	struct dac_channel_cfg cfg = {
+		.channel_id = 0,
+		.resolution = 8,
+	};
+	int ret = dac_channel_setup(dac_emul0, &cfg);
+
+	zassert_equal(ret, 0, "Setup should succeed");
+
+	/* Try to write value exceeding 8-bit range */
+	ret = dac_write_value(dac_emul0, 0, 256);
+	zassert_equal(ret, -EINVAL, "Out of range value should return -EINVAL");
+
+	ret = dac_write_value(dac_emul0, 0, 1000);
+	zassert_equal(ret, -EINVAL, "Out of range value should return -EINVAL");
+}
+
+ZTEST(dac_emul_tests, test_write_value_unconfigured_channel)
+{
+	/* Try to write to unconfigured channel */
+	int ret = dac_write_value(dac_emul_unconfigured, 0, 100);
+
+	zassert_equal(ret, -ENXIO, "Write to unconfigured channel should return -ENXIO");
+}
+
+ZTEST(dac_emul_tests, test_write_value_invalid_channel)
+{
+	/* Try to write to non-existent channel */
+	int ret = dac_write_value(dac_emul0, 10, 100);
+
+	zassert_equal(ret, -EINVAL, "Write to invalid channel should return -EINVAL");
+}
+
+/*** Read Value Tests ***/
+
+ZTEST(dac_emul_tests, test_read_value_basic)
+{
+	/* Setup and write */
+	struct dac_channel_cfg cfg = {
+		.channel_id = 0,
+		.resolution = 10,
+	};
+	int ret = dac_channel_setup(dac_emul0, &cfg);
+
+	zassert_equal(ret, 0, "Setup should succeed");
+
+	ret = dac_write_value(dac_emul0, 0, 512);
+	zassert_equal(ret, 0, "Write should succeed");
+
+	/* Read back */
+	uint32_t value;
+
+	ret = dac_emul_value_get(dac_emul0, 0, &value);
+	zassert_equal(ret, 0, "Read should succeed");
+	zassert_equal(value, 512, "Read value should match");
+}
```

Comment:
As far as I can see, this does exactly the same as `test_write_value_basic`. I suggest to keep only one of the two.